proof/refine/ARM/KHeap_R.thy

```diff
@@ -74,65 +74,187 @@ abbreviation (input)
-  shows      "\<lbrace>P\<rbrace> getObject p \<lbrace>\<lambda>(rv :: 'a :: pspace_storable). P\<rbrace>"
-  by (simp add: getObject_def split_def | wp x)+
+  "\<lbrace>P\<rbrace> getObject p \<lbrace>\<lambda>(rv :: 'a :: pspace_storable). P\<rbrace>"
+  unfolding getObject_def by wpsimp
 
 lemma getObject_tcb_inv [wp]: "\<lbrace>P\<rbrace> getObject l \<lbrace>\<lambda>(rv :: Structures_H.tcb). P\<rbrace>"
-  apply (rule getObject_inv)
-  apply simp
-  apply (rule loadObject_default_inv)
-  done
-
-lemma no_fail_loadObject_default [wp]:
-  "no_fail (\<lambda>s. \<exists>obj. projectKO_opt ko = Some (obj::'a) \<and>
+  by (rule getObject_inv)
+
+lemma loadObject_default_Some [simp]:
+  "\<lbrakk>projectKO_opt ko = Some (obj::'a);
+                      is_aligned p (objBits obj);
+                      case_option True (\<lambda>x. 2 ^ (objBits obj) \<le> x - p) n; q = p\<rbrakk>
+       \<Longrightarrow> bound (loadObject_default p q n ko s:: ('a::pre_storable) option)"
+  by (clarsimp simp: loadObject_default_def split_def projectKO_def obind_def
+                        alignCheck_def alignError_def magnitudeCheck_def projectKOs
+                        read_alignCheck_def read_alignError_def read_magnitudeCheck_def
+                        unless_def gets_the_def is_aligned_mask in_omonad
+             split: option.splits) simp
+
+lemmas loadObject_default_Some'[simp, intro!] = loadObject_default_Some[simplified]
+lemmas loadObject_default_Some''[simp, intro!]
+        = loadObject_default_Some[where p=p and s=s and n="snd (lookupAround2 p (ksPSpace s))" for p s,
+                                 simplified]
+
+lemma no_fail_r_loadObject_default [simp]:
+  "no_fail_r (\<lambda>s. \<exists>obj. projectKO_opt ko = Some (obj::'a) \<and>
                       is_aligned p (objBits obj) \<and> q = p
                       \<and> case_option True (\<lambda>x. 2 ^ (objBits obj) \<le> x - p) n)
-           (loadObject_default p q n ko :: ('a::pre_storable) kernel)"
-  apply (simp add: loadObject_default_def split_def projectKO_def
-                   alignCheck_def alignError_def magnitudeCheck_def
-                   unless_def)
-  apply (rule no_fail_pre)
-   apply (wp case_option_wp)
-  apply (clarsimp simp: is_aligned_mask)
-  apply (clarsimp split: option.split_asm)
-  apply (clarsimp simp: is_aligned_mask[symmetric])
-  apply simp
-  done
-
-lemma no_fail_getObject_tcb [wp]:
+           (loadObject_default p q n ko :: ('a::pre_storable) kernel_r)"
+  by (clarsimp simp: no_fail_r_def)
+
+method no_fail_r_readObject_method =
+  clarsimp simp: obj_at'_def readObject_def obind_def in_omonad split_def projectKO_eq no_fail_r_def,
+  rule ps_clear_lookupAround2, assumption+, simp,
+  blast intro: is_aligned_no_overflow,
+  clarsimp simp: objBits_simps' project_inject split: option.splits
+
+lemma no_fail_r_obj_at'_readObject_tcb[simp]:
+  "no_fail_r (obj_at' (P::tcb \<Rightarrow> bool) p) (readObject p::tcb kernel_r)"
```

Comment:
I can never quite decide if it is better to have `P` here or `\top`. `P` matches more often, but if used as intro rule will leave schematics in the final goal if the goal is produced `wp`-style. I think we may want `P` with `[simp]`, and `\top` with `intro!`. @mbrcknl and @Xaphiosis what is your take on these?

Reply:
That's a reasonable assessment. So keep this lemma as is, but potentially add a second one?
```
lemmas no_fail_r_obj_at'_readObject_tcb_top[wp]
  = no_fail_r_obj_at'_readObject_tcb[where P="\top"]
```
The alternative would be to only have the rule with `\top`, but then you might need to add weakening rules for both `no_fail_r` and `obj_at'` to get it to work with `simp`. I don't think you want those weakening rules in the default `simp` set.

Reply:
actually, "\top” versions are listed further below. I will make them `wp` rules rather than simp.